lib/qos-packet.js

```diff
@@ -18,6 +18,14 @@ function QoSPacket (original, client) {
   } else {
     this.messageId = original.messageId
   }
+
+  this.setDupIfNecessary = (persistence, client, cb) => {
+    if (this.qos === 2) {
+      persistence.outgoingUpdate(client, { ...this, dup: true }, cb)
+    } else {
+      cb()
+    }
+  }
```

Comment:
Move this to a top level function, allocating one for each packet is too expensive.

Reply:
Done.